### PR TITLE
Uptime command

### PR DESCRIPTION
**Pull request by Innectic**
Add uptime command

This needs to become service specific. This needs #60

### CODE REVIEW

**Comment by Innectic** [comment]
For some reason my stashed changes got in here :/

**Comment by Innectic** [comment]
Much better

File: cactusbot/api.py [hunk 1]
```diff
@@ -13,7 +13,7 @@ def __init__(self, channel, **kwargs):
 
         self.channel = channel
 
-    async def add_command(self, name, response, *, modifiers={},
+    async def add_command(self, name, response, *, permissions={},
```
Comment (2Cubed) [inline cactusbot/api.py]: Defaults should *never* be containers.
```python
In [1]: def f(x, a=[]):
   ...:     a.append(x)
   ...:     return a
   ...: 

In [2]: f(1)
Out[2]: [1]

In [3]: f(2)
Out[3]: [1, 2]

In [4]: f(3)
Out[4]: [1, 2, 3]
```
Instead, set it to `None`, and use `if permissions is None:` to reset it.

File: cactusbot/commands/magic/command.py [hunk 2]
```diff
@@ -15,7 +15,7 @@ class Meta(Command):
     }
 
     @Command.subcommand
-    async def add(self, command: r'!?([+$]?)(.+)', *response,
+    async def add(self, command: r'!?([+$%]?)(.+)', *response,
```
Comment (2Cubed) [inline cactusbot/commands/magic/command.py]: The `%` shouldn't be there *yet* - will break things.

File: cactusbot/commands/magic/uptime.py [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+from . import Command
+
+import aiohttp
+
+
+class Uptime(Command):
+    """Uptime command."""
+
+    @Command.subcommand()
+    async def get(self, *args: False, channel: "channel"):
```
Comment (2Cubed) [inline cactusbot/commands/magic/uptime.py]: No need for `*args` at all. :smile:

File: cactusbot/commands/magic/uptime.py [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+        data = await (await aiohttp.get(
+            "https://beam.pro/api/v1/channels/160788/manifest.light2".format(
+                channel))).json()
+
+        if data.get("startedAt") is not None:
```
Comment (2Cubed) [inline cactusbot/commands/magic/uptime.py]: `if "startedAt" in data` would be cleaner.

File: cactusbot/commands/magic/uptime.py [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+import aiohttp
+
+
+class Uptime(Command):
+    """Uptime command."""
+
+    @Command.subcommand()
+    async def get(self, *args: False, channel: "channel"):
+        data = await (await aiohttp.get(
+            "https://beam.pro/api/v1/channels/160788/manifest.light2".format(
```
Comment (2Cubed) [inline cactusbot/commands/magic/uptime.py]: Hardcoded channel! :cry:

File: cactusbot/commands/magic/uptime.py [hunk 6]
```diff
@@ -0,0 +1,30 @@
...
+                channel))).json()
+
+        if data.get("startedAt") is not None:
+            return "Channel has been live for {}.".format(
+                re.match(
+                    r"(.+)\.\d{6}",
+                    str(datetime.utcnow() - datetime.strptime(
+                        data["startedAt"][:-5], "%Y-%m-%dT%H:%M:%S"))
```
Comment (2Cubed) [inline cactusbot/commands/magic/uptime.py]: Use `"%Y-%m-%dT%H:%M:%S.%fZ"`, rather than `[:-5]`.

File: cactusbot/commands/magic/uptime.py [hunk 7]
```diff
@@ -0,0 +1,30 @@
...
+    @Command.subcommand()
+    async def get(self, *args: False, channel: "channel"):
+        data = await (await aiohttp.get(
+            "https://beam.pro/api/v1/channels/160788/manifest.light2".format(
+                channel))).json()
+
+        if data.get("startedAt") is not None:
+            return "Channel has been live for {}.".format(
+                re.match(
+                    r"(.+)\.\d{6}",
```
Comment (2Cubed) [inline cactusbot/commands/magic/uptime.py]: Cleaner without regex here. If you want to ignore microseconds:
```python
time = datetime.utcnow() - datetime.strptime(...)

# subtract microseconds, zero microseconds will not be displayed in the str()
time -= datetime.timedelta(microseconds=time.microseconds)

print(str(time))
```

File: cactusbot/commands/magic/command.py (hunk 2 above)
Comment (Innectic) [inline cactusbot/commands/magic/command.py]: Oh yeah, forgot to remove that from my stash-removing-commit

**Pull request merged**